File: backend/app/schemas/insights.py
"""Schemas for insights endpoint responses"""

import uuid
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.fx import FxStatus

NetWorthGroupKind = Literal["asset", "debt"]
InsightsComparisonPeriod = Literal["same_length", "previous_month", "previous_year"]
SavedInsightsRangeUnit = Literal["day", "week", "month", "quarter", "year"]

# this = current period to date, last = previous complete periods, past = rolling window
SavedInsightsRangeQualifier = Literal["this", "last", "past"]

# Guards a saved range against absurd look-backs while leaving room for any sensible window
SAVED_INSIGHTS_RANGE_MAX_AMOUNT = 999


class InsightsPeriodAtAGlanceResponse(BaseModel):
    """Compact payload for the insights Period At A Glance card"""

    income: int
    expenses: int
    income_expense_fx_status: FxStatus = Field(default_factory=FxStatus)
    net_worth_change: int
    net_worth_change_fx_status: FxStatus = Field(default_factory=FxStatus)
    top_category_name: str | None = None
    top_category_share_pct: int | None = None
    top_category_fx_status: FxStatus = Field(default_factory=FxStatus)
    biggest_change_name: str | None = None
    biggest_change_amount: int | None = None
    biggest_change_pct: int | None = None
    biggest_change_fx_status: FxStatus = Field(default_factory=FxStatus)


class InsightsFundFlowResponse(BaseModel):
    """Payload for the insights Fund Flow card."""

    income_sources: list[tuple[str, int]]
    expense_categories: list[tuple[str, int]]
    income_outflows: list[tuple[str, int]]
    expense_inflows: list[tuple[str, int]]
    income_source_count: int
    expense_category_count: int
    fx_status: FxStatus = Field(default_factory=FxStatus)


class InsightsIncomeExpenseBreakdownResponse(BaseModel):
    """Payload for the insights income/expense breakdown card

    Pie rows are ``(id, name, original_category_kind, amount)``
    ``expense_total`` and ``income_total`` are authoritative center totals
    after flipped refund/loss categories are netted against their original side
    """

    expense: list[tuple[str, str, str, int]]
    income: list[tuple[str, str, str, int]]
    expense_total: int
    income_total: int
    expense_increases: list[tuple[str, str, int, int, int | None, int]]
    expense_decreases: list[tuple[str, str, int, int, int | None, int]]
    income_increases: list[tuple[str, str, int, int, int | None, int]]
    income_decreases: list[tuple[str, str, int, int, int | None, int]]
    fx_status: FxStatus = Field(default_factory=FxStatus)


class InsightsCashFlowResponse(BaseModel):
    """Payload for the insights cash-flow card."""

    points: list[tuple[date, date, int, int]]
    fx_status: FxStatus = Field(default_factory=FxStatus)


class InsightsNetWorthResponse(BaseModel):
    """Payload for the insights net-worth card."""

    groups: list[tuple[str, str, NetWorthGroupKind]]
    baseline: list[int] = Field(default_factory=list)
    points: list[tuple[date, date, list[int]]]
    fx_status: FxStatus = Field(default_factory=FxStatus)


class InsightsSavingsRateTrendResponse(BaseModel):
    """Payload for the insights savings-rate trend card."""

    points: list[tuple[date, int, int]]
    fx_status: FxStatus = Field(default_factory=FxStatus)


class InsightsMerchantDistributionResponse(BaseModel):
    """Payload for the insights merchant distribution card."""

    merchants: list[tuple[str, str, int, int | None, int | None]]


class InsightsMerchantRankingResponse(BaseModel):
    """Payload for the insights merchant ranking card."""

    merchants: list[tuple[str, str, int, int, int | None]]


class InsightsMerchantsResponse(BaseModel):
    """Shared payload for insights merchant cards."""

    distribution: list[tuple[str, str, int, int | None, int | None]]
    ranking: list[tuple[str, str, int, int, int | None]]
    fx_status: FxStatus = Field(default_factory=FxStatus)


class SavedInsightsRangeCreate(BaseModel):
    """Payload for saving a named relative insights range"""

    name: str = Field(min_length=1, max_length=64)
    amount: int = Field(ge=1, le=SAVED_INSIGHTS_RANGE_MAX_AMOUNT)
    unit: SavedInsightsRangeUnit
    qualifier: SavedInsightsRangeQualifier = "past"


class SavedInsightsRangeResponse(BaseModel):
    """A user's saved relative insights range"""

    id: uuid.UUID
    name: str
    amount: int
    unit: SavedInsightsRangeUnit
    qualifier: SavedInsightsRangeQualifier
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/api/insights.ts
import { useQuery } from '@tanstack/react-query';
import { useAuth } from '@/hooks/useAuth';
import { authenticatedFetch } from '@/api/client';
import type { FxStatus } from '@/api/dashboard';
import { insightsKeys } from '@/api/queryKeys';
import { buildQueryString } from '@/api/queryString';
import type { InsightsComparisonPeriod } from '@/insights/types/range';

export interface InsightsPeriodGlanceResponse {
  income: number;
  expenses: number;
  income_expense_fx_status: FxStatus;
  net_worth_change: number;
  net_worth_change_fx_status: FxStatus;
  top_category_name?: string;
  top_category_share_pct?: number;
  top_category_fx_status: FxStatus;
  biggest_change_name?: string;
  biggest_change_amount?: number;
  biggest_change_pct?: number;
  biggest_change_fx_status: FxStatus;
}

export type InsightsFlowEntry = [string, number];

export interface InsightsFundFlowResponse {
  income_sources: InsightsFlowEntry[];
  expense_categories: InsightsFlowEntry[];
  income_outflows: InsightsFlowEntry[];
  expense_inflows: InsightsFlowEntry[];
  income_source_count: number;
  expense_category_count: number;
  fx_status: FxStatus;
}

export type InsightsBreakdownCategoryKind = 'expense' | 'income';
export type InsightsBreakdownEntry = [string, string, InsightsBreakdownCategoryKind, number];
export type InsightsCategoryTrendEntry = [string, string, number, number, number | null, number];

export interface InsightsIncomeExpenseBreakdownResponse {
  expense: InsightsBreakdownEntry[];
  income: InsightsBreakdownEntry[];
  expense_total: number;
  income_total: number;
  expense_increases: InsightsCategoryTrendEntry[];
  expense_decreases: InsightsCategoryTrendEntry[];
  income_increases: InsightsCategoryTrendEntry[];
  income_decreases: InsightsCategoryTrendEntry[];
  fx_status: FxStatus;
}

export type InsightsCashFlowPoint = [string, string, number, number];

export interface InsightsCashFlowResponse {
  points: InsightsCashFlowPoint[];
  fx_status: FxStatus;
}

export type InsightsNetWorthGroup = [string, string, 'asset' | 'debt'];
export type InsightsNetWorthPoint = [string, string, number[]];

export interface InsightsNetWorthResponse {
  groups: InsightsNetWorthGroup[];
  baseline: number[];
  points: InsightsNetWorthPoint[];
  fx_status: FxStatus;
}

export type InsightsSavingsRateTrendPoint = [string, number, number];

export interface InsightsSavingsRateTrendResponse {
  points: InsightsSavingsRateTrendPoint[];
  fx_status: FxStatus;
}

export type InsightsMerchantDistributionEntry = [string, string, number, number | null, number | null];

export type InsightsMerchantRankingEntry = [string, string, number, number, number | null];

export interface InsightsMerchantsResponse {
  distribution: InsightsMerchantDistributionEntry[];
  ranking: InsightsMerchantRankingEntry[];
  fx_status: FxStatus;
}

/**
 * Builds the backend range query shared by dated insights widgets
 */
function buildInsightsRangeQueryString(
  fromDate: string,
  toDate: string,
  comparisonPeriod?: InsightsComparisonPeriod,
) {
  return buildQueryString({
    from_date: fromDate,
    to_date: toDate,
    comparison_period: comparisonPeriod,
  });
}

/**
 * Fetches headline metrics for an insights date range
 */
export function fetchInsightsPeriodGlance(
  fromDate: string,
  toDate: string,
  comparisonPeriod: InsightsComparisonPeriod = 'same_length',
) {
  return authenticatedFetch<InsightsPeriodGlanceResponse>(
    `/insights/period-glance${buildInsightsRangeQueryString(fromDate, toDate, comparisonPeriod)}`,
  );
}

export function useInsightsPeriodGlance(
  fromDate: string,
  toDate: string,
  comparisonPeriod: InsightsComparisonPeriod = 'same_length',
  enabled = true,
) {
  const { accessToken } = useAuth();
  return useQuery({
    queryKey: insightsKeys.periodGlance(fromDate, toDate, comparisonPeriod),
    queryFn: () => fetchInsightsPeriodGlance(fromDate, toDate, comparisonPeriod),
    enabled: !!accessToken && enabled && fromDate !== '' && toDate !== '',
    staleTime: 5 * 60 * 1000,
  });
}

/**
 * Fetches net worth groups and points for an insights date range
 */
export function fetchInsightsNetWorth(fromDate: string, toDate: string) {
  return authenticatedFetch<InsightsNetWorthResponse>(
    `/insights/net-worth${buildInsightsRangeQueryString(fromDate, toDate)}`,
  );
}

export function useInsightsNetWorth(fromDate: string, toDate: string, enabled = true) {
  const { accessToken } = useAuth();
  return useQuery({
    queryKey: insightsKeys.netWorth(fromDate, toDate),
    queryFn: () => fetchInsightsNetWorth(fromDate, toDate),
    enabled: !!accessToken && enabled && fromDate !== '' && toDate !== '',
    staleTime: 5 * 60 * 1000,
  });
}

/**
 * Fetches the global savings-rate trend used by the insights page
 */
export function fetchInsightsSavingsRateTrend() {
  return authenticatedFetch<InsightsSavingsRateTrendResponse>('/insights/savings-rate-trend');
}

export function useInsightsSavingsRateTrend(enabled = true) {
  const { accessToken } = useAuth();
  return useQuery({
    queryKey: insightsKeys.savingsRateTrend(),
    queryFn: fetchInsightsSavingsRateTrend,
    enabled: !!accessToken && enabled,
    staleTime: 5 * 60 * 1000,
  });
}

/**
 * Fetches income and expense category movement for an insights date range
 */
export function fetchInsightsIncomeExpenseBreakdown(
  fromDate: string,
  toDate: string,
  comparisonPeriod: InsightsComparisonPeriod = 'same_length',
) {
  return authenticatedFetch<InsightsIncomeExpenseBreakdownResponse>(
    `/insights/income-expense-breakdown${buildInsightsRangeQueryString(
      fromDate,
      toDate,
      comparisonPeriod,
    )}`,
  );
}

export function useInsightsIncomeExpenseBreakdown(
  fromDate: string,
  toDate: string,
  comparisonPeriod: InsightsComparisonPeriod = 'same_length',
  enabled = true,
) {
  const { accessToken } = useAuth();
  return useQuery({
    queryKey: insightsKeys.incomeExpenseBreakdown(fromDate, toDate, comparisonPeriod),
    queryFn: () => fetchInsightsIncomeExpenseBreakdown(fromDate, toDate, comparisonPeriod),
    enabled: !!accessToken && enabled && fromDate !== '' && toDate !== '',
    staleTime: 5 * 60 * 1000,
  });
}

/**
 * Fetches cash-flow points for an insights date range
 */
export function fetchInsightsCashFlow(fromDate: string, toDate: string) {
  return authenticatedFetch<InsightsCashFlowResponse>(
    `/insights/cash-flow${buildInsightsRangeQueryString(fromDate, toDate)}`,
  );
}

export function useInsightsCashFlow(fromDate: string, toDate: string, enabled = true) {
  const { accessToken } = useAuth();
  return useQuery({
    queryKey: insightsKeys.cashFlow(fromDate, toDate),
    queryFn: () => fetchInsightsCashFlow(fromDate, toDate),
    enabled: !!accessToken && enabled && fromDate !== '' && toDate !== '',
    staleTime: 5 * 60 * 1000,
  });
}

/**
 * Fetches income and expense flow groups for an insights date range
 */
export function fetchInsightsFundFlow(fromDate: string, toDate: string) {
  return authenticatedFetch<InsightsFundFlowResponse>(
    `/insights/fund-flow${buildInsightsRangeQueryString(fromDate, toDate)}`,
  );
}

export function useInsightsFundFlow(fromDate: string, toDate: string, enabled = true) {
  const { accessToken } = useAuth();
  return useQuery({
    queryKey: insightsKeys.fundFlow(fromDate, toDate),
    queryFn: () => fetchInsightsFundFlow(fromDate, toDate),
    enabled: !!accessToken && enabled && fromDate !== '' && toDate !== '',
    staleTime: 5 * 60 * 1000,
  });
}

/**
 * Fetches merchant distribution and ranking data for an insights date range
 */
export function fetchInsightsMerchants(
  fromDate: string,
  toDate: string,
  comparisonPeriod: InsightsComparisonPeriod = 'same_length',
) {
  return authenticatedFetch<InsightsMerchantsResponse>(
    `/insights/merchants${buildInsightsRangeQueryString(fromDate, toDate, comparisonPeriod)}`,
  );
}

export function useInsightsMerchants(
  fromDate: string,
  toDate: string,
  comparisonPeriod: InsightsComparisonPeriod = 'same_length',
  enabled = true,
) {
  const { accessToken } = useAuth();
  return useQuery({
    queryKey: insightsKeys.merchants(fromDate, toDate, comparisonPeriod),
    queryFn: () => fetchInsightsMerchants(fromDate, toDate, comparisonPeriod),
    enabled: !!accessToken && enabled && fromDate !== '' && toDate !== '',
    staleTime: 5 * 60 * 1000,
  });
}
